### DISCUSSION TITLE
Dash frontend

### DISCUSSION THREAD

**Discussion by Anonymous5726221** [Anonymous5726221/lapalma_data#7] [Ideas]
Currently charts are displayed side by side in 2 columns. 

![image](https://user-images.githubusercontent.com/78229427/137604151-908b276b-d8b5-4fd8-af7f-bcb1a2235780.png)

The idea is to be able to see multiple charts at once to get the "bigger picture". The issue is that this significantly reduce readability of the charts.
We need to find a solution which would allow to keep this arrangement but give the option to see charts in a bigger format when needed. And if possible without overloading the website.

One such solution could be [dash-bootstrap-components](https://dash-bootstrap-components.opensource.faculty.ai) which offers a lot more front end option than native dash.

I found this [post](https://community.plotly.com/t/dash-graph-display-on-full-screen/11118/5) on plotly forum which demonstrate a way to add a button to charts to make them full screen which would be a great option. Unfortunately this post quality is very poor and doesn't give much information other than it is possible to do it.

**Comment by ontracrous** [comment]
I was playing around with dash and bootstrap to see if I could come up with a layout that makes things easy to read and navigate through.
Currently got a base layout that looks like this:
![image](https://user-images.githubusercontent.com/80160339/137606099-4138fa05-0c0d-482b-933a-96568f6f7fe4.png)

My idea was to not overwhelm people with data when they look at it and categorize the data.
Each view would consist of:
- A single plot;
- grouped plots.
That way people don't get confused when they look at the data.

With the help of tabs and dropdown menus people could switch in between different views like I did with the picture above, switching between statistics.

If you like the idea I can try and setup a framework for the layout so we can implement the plots

**Reply by Anonymous5726221** [reply]
I think that looks good. I'd still want to keep all of them as well somewhere though, probably on the home page. The `today`, `daily`, `weekly` and `total` tabs are a good idea. And I like the graphs categories.

A few thoughts:
- The `today`, `daily`, `weekly` and `total` tabs could persist on the other views. 
  - We could use them as preset for the existing date filter. 
  - We could also add a new filters to change the groupby behavior
- On the other views how do we want to display the graphs?
  - Simply the graph in full page and simply scrolling to the next (I think we should go with that first, should be easy to implement)
  - One graph in full page and a ribbon below it with the list of graphs available?
  - Some kind of other chart selector

Feel free to create a draft PR so I can have a look as well and maybe I can take care of some part of it if you want.

**Reply by ontracrous** [reply]
I'm currently in the works of setting up a framework that allows us to quickly change things around without having to scroll through giant piles of code. And create a little more structure.
![image](https://user-images.githubusercontent.com/80160339/137630055-d79ce831-05d0-4fab-ae5c-72a6955f61f1.png)
This is currently in a separate project just so I don't have any interference from any existing callbacks and whatnot. Once I'm done with the framework I'll add the existing software in it and PR it.

I like the idea of having the tabs persist for every view, however we do need to change the plots around a little bit if that were the case.
We currently don't support `daily`, `weekly` and `total` in the plots at the moment.
And if we do change things, I would suggest having a `monthly` tab as well. Also because we're currently +8 weeks in,

We're going to be free on which view shows what graphs and how. If we for whatever reason want the map plot and the histogram plot we can simply put that in the `viewModules` as a `html.Div()` and depending on the selected view in `sidebar` it displays only those graphs.

We could also have another view which has a dropdown menu and a `add graph` and `remove graph` button of some sorts. Which allows the user to add different graphs to create their own view. That's just an idea and a "nice to have" feature.

**Comment by ontracrous** [comment]
I finished the framework, got a demo here [Dash_Framework repository](https://github.com/ontracrous/Dash_Framework)
There are some serious changes coming with this, so before I rework pretty much everything I wanna know your opinion.

Hope you like the way things are handled and how much easier it is to use/debug

**Reply by ontracrous** [reply]
@Anonymous5726221 Forgot to @ you.

**Reply by Anonymous5726221** [reply]
That's beautiful, I really like it!!

![image](https://user-images.githubusercontent.com/78229427/137639916-df23daba-bdac-44e3-9f38-294afb9addd1.png)

And the code structure is much tidier. The settings on the left, common to each charts on a view is a great idea.
I'm definitely ok with using that.

At the moment I'm using a janky solution using a `dcc.interval` callback to fetch the data from the database every 10 seconds with a cache timeout of 1 minute (stupid solution but it works), can this be improved with this setup? At worst can it be replicated?

Awesome work fren, excited to have that setup. 

How do you want to proceed? Do you want to split the work or you want to do it all yourself.

**Reply by Anonymous5726221** [reply]
Once that's implemented a few things we should think of are:
- Have the graph style somewhat match the page style.
  - Possibly a style defined in a different file that can be passed around to the different charts.
- Think of a ways to handle those `daily`, `weekly`, ... tabs
  - Do we want to keep the same graph but with data for the day, week, month, ...
  - Do we want to change the groupy in the figure (e.g. instead of showing mean magnitude for a day, show for a week). <-- That's the one I prefer
- Have subsections for each categories (Magnitude, Energy, ...) to allow displaying each chart individually in full page.

**Reply by ontracrous** [reply]
> At the moment I'm using a janky solution using a dcc.interval callback to fetch the data from the database every 10 seconds with a cache timeout of 1 minute (stupid solution but it works), can this be improved with this setup? At worst can it be replicated?

The way I have the framework implemented right now does calls to a database to get information. It doesn't do anything else with the database, so we're likely going to have to implement your workaround. I'll have a look at that, see if we can somehow improve on that.

>How do you want to proceed? Do you want to split the work or you want to do it all yourself.

I'll do most of the refactoring to the new layout. Some database stuff I'll leave to you, because I don't know the specifics of that.

>Have the graph style somewhat match the page style.

I have an `assets` folder for .css files, but for some reason it didn't work... I still gotta look into that as to why I'm not allowed to load those files.

>Think of a ways to handle those daily, weekly, ... tabs

Might change the tabs to be `pills` in the settings section to keep things more uniform.
It's easier that way in terms of usability.

>Do we want to keep the same graph but with data for the day, week, month,

I'm not entirely sure yet what we should do with the settings.
What I'm currently thinking of is the following:
- select anything other than `today`
- a button/arrow is available to move back- or forwards through the data

So you can go through all the stats simply by clicking on those buttons.
And that way it's scalable for whatever filter is selected

>Do we want to change the groupy in the figure (e.g. instead of showing mean magnitude for a day, show for a week). <-- That's the one I prefer

We could let the user decide what they want to see. You can mess around with "color" in plots to get different results
Like if you use `magnitude` as color you get the colorbar, if you change it to the group thing you made it goes it has (de)selectable items in the plot.
We could do something like that.

First thing we should do is have the new new software up and running and afterwards we can decide how to proceed.


Let me know what you think of the idea I have for the `daily`, `weekly`... combined with arrows to scroll through the data.

**Reply by ontracrous** [reply]
Just thought of a possible solution for the data gathering!

What is we created 2 threads. One thread for the server another thread that collects data every minute or whatever.
Thread1 will collect data with the Seismicportal, puts it in the database and that's all it does.
Thread2 will be the website that calls on the database to get the info it needs to create plots and other stuff.

That is the cleanest I can come up with, other than having 2 separate .py scripts running instead of 2 threads.

**Reply by ontracrous** [reply]
Decided to change the `settings` part of the sidebar to show you what I meant with scrolling through data
![image](https://user-images.githubusercontent.com/80160339/137645447-4e3b9521-7c5b-406e-b061-74e2acb8a07d.png)

@Anonymous5726221

**Reply by Anonymous5726221** [reply]
- I like today/daily/weekly/... as `pills`
- Not sure about the arrows to move through the data, I like the date range selector better I think.
- This app only `GET` data from the database, I've got a separate service running elsewhere updating the database live with events coming from websocket. There is probably a better way to `GET` from a database with Flask. Dash connection to database is only for enterprise use. I need to look into all that.

**Reply by ontracrous** [reply]
How would you incorporate the today/daily/etc. filters with a date picker?
I don't really see a way to have it work for weekly and monthly statistics.
`today` and `total` can just be fixed dates, but the other ones are not as intuitive I feel.

**Comment by Anonymous5726221** [comment]
@ontracrous Merging was actually much simpler than I thought it would be a tons of change but only a minor conflict in the end.

I'm gonna do a final test and I should be able to push to heroku in the next few minutes.

:confetti_ball: :tada: Excellent work on the UI redesign. :tada: :confetti_ball:

**Reply by ontracrous** [reply]
Glad to hear it wasn't that bad!
Excited to see it in action!

**Reply by Anonymous5726221** [reply]
I'm investigating an issue but I'm almost done.
